### PR TITLE
GSoC Add ONNX Support for GatherElements

### PR DESCRIPTION
**Pull request by Aser-Abdelfatah**
Merge with https://github.com/opencv/opencv/pull/24092

This pull request serves as an addition to the pull request "GSoC Add ONNX Support for GatherElements #24092", which is submitted to the original OpenCV repository. It contains three tests to the GatherElements operator. 
The three tests are sourced from:
* https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node/test_gather_elements_0
* https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node/test_gather_elements_1
* https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node/test_gather_elements_negative_indices
And the models were configured using the code provided by Yuantao Feng at https://github.com/fengyuentau/onnx_utils to confirm output shape in pre-run.

---

Reference outputs for `optical_flow_estimation_raft_2023aug.onnx` is generated with the following script:

```python
import cv2 as cv
import numpy as np

img1 = cv.imread("./opencv_extra/testdata/gpu/opticalflow/frame0.png")
img2 = cv.imread("./opencv_extra/testdata/gpu/opticalflow/frame1.png")

img1_blob = cv.cvtColor(img1, cv.COLOR_BGR2RGB)
img1_blob = cv.resize(img1_blob, (480, 360))
# source: (1, 3, 388, 584)
# target: 1, 3, 360, 480
img1_blob = cv.dnn.blobFromImage(img1_blob)
# source:  (1, 3, 388, 584)
# target: 1, 3, 360, 480
img2_blob = cv.cvtColor(img2, cv.COLOR_BGR2RGB)
img2_blob = cv.resize(img2_blob, (480, 360))
img2_blob = cv.dnn.blobFromImage(img2_blob)
print(img1_blob.shape, img2_blob.shape)

import onnxruntime as ort
net = ort.InferenceSession("optical_flow_estimation_raft_2023aug.onnx", providers=["CPUExecutionProvider"])
output = net.run(["12007", "12006"], {"0": img1_blob, "1": img2_blob})
np.save("output_optical_flow_estimation_raft_2023aug_0.npy", output[0])
np.save("output_optical_flow_estimation_raft_2023aug_1.npy", output[1])
print(output[0].shape)
print(output[1].shape)
```

### CODE REVIEW

File: testdata/dnn/download_models.py [hunk 1]
```diff
@@ -178,6 +178,11 @@ def get_confirm_token(response):  # in case of large files
 
 
 models = [
+    Model(
+        name='RAFT',
+        url='https://drive.google.com/u/0/uc?id=1yWIbF4birkpZMhNNhqzy75d4EvJW-UXE&export=download',
```
Comment (dkurt) [inline testdata/dnn/download_models.py:183]: Please share a source of that link in comment (i.e. origin repository)

Comment (dkurt) [inline testdata/dnn/download_models.py:183]: Use `GDrive` class for Google Drive models:
```
downloader=GDrive('1yWIbF4birkpZMhNNhqzy75d4EvJW-UXE'),
```

File: testdata/dnn/onnx/data/input_RAFT_0.npy
Comment (dkurt) [inline testdata/dnn/onnx/data/input_RAFT_0.npy:1]: Too big file. What is the data type and dimensions?

**Review by dkurt — CHANGES_REQUESTED** [review]
`input_RAFT_0.npy` and `input_RAFT_1.npy` sizes too big

File: testdata/dnn/download_models.py (hunk 1 above)
Comment (Aser-Abdelfatah) [reply testdata/dnn/download_models.py:183]: The model is sourced from here https://github.com/Aser-Abdelfatah/opencv_zoo/tree/gsoc_RAFT_Model_ONNX/models/optical_flow_estimation_raft

**Comment by fengyuentau** [comment]
Could you use one of the existing image in this repo, such as https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/dog_orig_size.png? You can resize it to the target scale, run inference, then collect the corresponding output, so that the total size can be cut by half, which passes the limit check.

File: testdata/dnn/download_models.py (hunk 1 above)
Comment (dkurt) [reply testdata/dnn/download_models.py]: I mean it there an officially distributed model? There is a little chance that custom file will be available for a long time.

Comment (dkurt) [reply testdata/dnn/download_models.py]: I think it make sense to highlight all that details at least as comments before the Model block.

**Comment by Aser-Abdelfatah** [comment]
> Could you use one of the existing image in this repo, such as https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/dog_orig_size.png? You can resize it to the target scale, run inference, then collect the corresponding output, so that the total size can be cut by half, which passes the limit check.

Sorry for getting back late to this. Since it's an optical flow model, the input should be two images. I can use the same image twice for both inputs, but it wouldn't really make much sense. Another approach would be finding two lightweight images, but a question is what is the limit check?

**Comment by dkurt** [comment]
@Aser-Abdelfatah, you may use images from https://github.com/opencv/opencv_extra/tree/4.x/testdata/gpu/opticalflow. Preprocess them in runtime.

**Comment by Aser-Abdelfatah** [comment]
@dkurt I didn't get the part about pre-processing in runtime. The testing code in opencv/modules/dnn/test/test_onnx_importer.cpp is essentially, 

`TEST_P(Test_ONNX_layers, RAFT)
{
    testONNXModels("RAFT", npy, 0, 0, false, true, 2);
}`
So, the model expects two pre-processed inputs.
And of course, when I pre-process the images using a separate python code and convert the pre-processed numpy array to an npy file, the size of each input is the same, essentially 2.0 mega bytes.

**Comment by Aser-Abdelfatah** [comment]
And sorry for keeping getting back late to this. The time difference is crazy

**Comment by dkurt** [comment]
> And of course, when I pre-process the images using a separate python code and convert the pre-processed numpy array to an npy file

So you can do the same in C++ - unread two images and preprocess using OpenCV methods.

**Comment by Aser-Abdelfatah** [comment]
I'm struggling with the C++ implementation. Essentially, I can't find a way to export the 4-dimensional opencv Mat to npy file in C++.

**Comment by fengyuentau** [comment]
You can do it in Python with OpenCV and NumPy.

```python
import cv2 as cv
import numpy as np

img1 = cv.imread("/path/to/img1")
img2 = cv.imread("/path/to/img2")

import onnxruntime as ort
net = ort.InferenceSession("/path/to/raft/model")
output = net.run(["output_name"], {"input1_name": img1, "input2_name": img2})
np.save("output_file_name.npy", output[0])
```

Maybe you need to debug a little bit to run the code above, but it should be mostly correct.

**Comment by Aser-Abdelfatah** [comment]
I did it before. The problem is that the resulting npy file is also 2 mega bytes

**Comment by fengyuentau** [comment]
2MB is fine. IIRC, the limit is 3MB.

**Comment by dkurt** [comment]
> 2MB is fine. IIRC, the limit is 3MB.

@fengyuentau, this is not fine. The preprocessing might be done in C++ without any issues. Such 2MB will be added to git history forever.

> I'm struggling with the C++ implementation.

@Aser-Abdelfatah, share Python implementation for preprocessing. Show how .npy are saved so we can propose an alternative C++ impl.

**Comment by fengyuentau** [comment]
How is it possible to test model accuracy if the output reference is not preserved in this repository?

What I am saying is we don't need to check the exact correct results as what common users expect, we don't need to perform preprocessings to make results correct and meaningful. Just put two input images that has the input scale that the model requires, save outputs as a npy file, put the output file here, and thats it.

**Comment by dkurt** [comment]
There is no problem with output - it takes just 21Kb (see `testdata/dnn/onnx/data/output_RAFT.npy`). This is a reference for test. I propose the following:

1. Use two images in Python and native RAFT to get output reference. Keep only output in .npy file
2. Use the same images in C++ test with the same preprocessing which was done in Python.

No need to add preprocessed input .npy to this PR.

**Comment by dkurt** [comment]
Seems like there is a misunderstanding. Proposed solution by @fengyuentau at https://github.com/opencv/opencv_extra/pull/1082#issuecomment-1705875232 if correct. I don't understand what is `testdata/dnn/onnx/data/output_RAFT.npy` and why it's 21Kb but not 2MB as @Aser-Abdelfatah mentioned. Let's upload actual files first.

**Comment by dkurt** [comment]
@Aser-Abdelfatah, thanks! This is C++ function you need to use:
```cpp
Mat image1 = cv::imread("frame0.png");
Mat input1 = cv::dnn::blobFromImage(image1, 1.0, Size(input_width, input_height), Scalar(), true);
```

**Review by dkurt — APPROVED** [review]

**Pull request merged**